tribe-ext-disallow-past-starting-dates.php

```diff
@@ -0,0 +1,517 @@
+
+/**
+ * Known issues:
+ * 1)
+ * If the user's local time zone and the site's time zone are on different
+ * calendar days (such as UTC+0 and UTC+1 and UTC is at 23:00:00), the
+ * datepicker will display the user's today as selectable but it won't be
+ * selectable. This is correct because the site owner does not want the user to
+ * select their own today because it's the site's yesterday.
+ *
+ * This odd UX is due to the jQuery UI Datepicker library not supporting
+ * time zones and the timepicker being a separate input.
+ *
+ * 2)
+ * All "protection" against setting an invalid start date is in JavaScript--none
+ * in PHP.
+ */
+
+// Do not load unless Tribe Common is fully loaded and our class does not yet exist.
+if ( class_exists( 'Tribe__Extension' ) && ! class_exists( 'Tribe__Extension__Custom_Datepicker_Start_Date' ) ) {
+	/**
+	 * Extension main class, class begins loading on init() function.
+	 */
+	class Tribe__Extension__Custom_Datepicker_Start_Date extends Tribe__Extension {
+		/**
+		 * The script's handle.
+		 */
+		public $handle = 'tribe-ext-custom-datepicker-start-date';
+
+		/**
+		 * The array of keys/values to send to the script.
+		 *
+		 * @see wp_localize_script()
+		 *
+		 * @var array
+		 */
+		private $script_vars = array();
+
+		/**
+		 * The existing start date's midnight timestamp.
+		 *
+		 * This will remain NULL if this is a new (non-existing) event or if its
+		 * start date is TODAY.
+		 *
+		 * @see Tribe__Extension__Custom_Datepicker_Start_Date::get_min_date()
+		 *
+		 * @var null|int
+		 */
+		private $timestamp_existing_start_date = null;
+
+		/**
+		 * The timestamp of whichever is greater between Today and an existing
+		 * event's start time, if applicable.
+		 *
+		 * This will remain NULL if this is a new (non-existing) event or if its
+		 * start date is TODAY.
+		 *
+		 * @see Tribe__Extension__Custom_Datepicker_Start_Date::get_min_date()
+		 *
+		 * @var null|int
+		 */
+		private $timestamp_max_today_or_existing = null;
+
+		/**
+		 * The script's handle with underscores instead of hyphens.
+		 *
+		 * Used for building filter and action hook names and CSS class name.
+		 *
+		 * @return string
+		 */
+		public function get_handle_underscores() {
+			return str_replace( '-', '_', $this->handle );
+		}
+
+		/**
+		 * The input's CSS error class.
+		 *
+		 * @return string
+		 */
+		private function get_error_css_class() {
+			return $this->get_handle_underscores() . '_error';
+		}
+
+		/**
+		 * The WordPress capability required to be able to pick any date.
+		 *
+		 * @return string
+		 */
+		private function get_cap_allowed_any_date() {
+			/**
+			 * The capability required to have this script not load.
+			 *
+			 * @param string $capability The minimum capability to be allowed to
+			 *                           choose any start date.
+			 *
+			 * @return string
+			 */
+			return (string) apply_filters( $this->get_handle_underscores() . '_cap_allowed_any_start_date', 'manage_options' );
```

Comment:
Dynamic hooks can be useful but of course it's not immediately apparent what they resolve to exactly. In this case especially, that's irksome because the final hook name isn't going to change (unless I'm misreading). 

That is, unlike something along the lines of `save_post_$POSTTYPE` - where it's pretty apparent as to what will change and why - in this case it seems like the final hook will never really vary.

Reply:
all static now